### ISSUE TITLE
Bug in visualization.cpp?

### ISSUE THREAD

**Issue by suiyoubi** [dilevin/CSC417-a3-finite-elements-3d#16]
When I was debugging my skinny bunny, I found that there's a matrix product dimension not match assertion failed. I manged to locate the assertion failed at visualization.cpp at line 370 (in func `Visualize::update_vertex_positions`), where here we are trying to product the "big matrix" N with dimen (~14000, 1500) with `g_geometry[g_id[id]].first` with dimen (500, 3).

If I understand correctly, we should mutiply N with a vector instead of a matrix, so my guess is we should use argument parameter `pos` instead of `g_geometry[g_id[id]].first`, and then reshape the product result as ( ~14000/3, 3). My fix is provided below and it fixed the problem (I can finally see the happy bunny when pressing 'S'):

```
...
...
if(g_skinning) {
            Eigen::VectorXd flat = g_N[id] * pos;
            Eigen::MatrixXd ans(flat.size() / 3, 3);
            for(int i = 0; i < flat.size()/3;i++) {
                ans(i, 0) = flat(3 * i);
                ans(i, 1) = flat(3 * i + 1);
                ans(i, 2) = flat(3 * i + 2);
            }
            g_viewer.data_list[g_id[id]].V = ans;
        } else
...
...
...
```
Since the last edit of this file is 13 months ago, I am not sure if I am doing the right thing here.. Could you please confirm that This is actually a BUG, or I probably did not interpret the N matrix correctly ?

**Comment by abhimadan** [comment]
I believe this is a bug in your skinning matrix. There was a mistake in the comment in `build_skinning_matrix.h` regarding the dimensions, so please check the updated file and make sure your dimensions are correct.

**Comment by suiyoubi** [comment]
> I believe this is a bug in your skinning matrix. There was a mistake in the comment in `build_skinning_matrix.h` regarding the dimensions, so please check the updated file and make sure your dimensions are correct.

Exactly, if N is ` 3l-by-3n`, then N must multiply by a `3n-by-1` vector, and what I wrote above is needed.
However, if N has the dimension 'l-by-n' (which was changed to `3l-by-3n` in commit f12a23abeba603e64d05545bece5e47e12a0ac3a), the dimension N can match everything in visualization.cpp.

I believe the `3l-by-3n` matrix is what is described in the video (Lecture5 24:07), but should we change it to `l-by-n` in order to fit what is hard-coded in visualization.cpp? (In other words, the comment was actually correct, and the commit is incorrect) ?

**Comment by abhimadan** [comment]
Ah yes you're right. I was going off a previous discussion but you're correct, the initial comment on that file was correct. Sorry about the confusion, I'll change that back.

**Comment by suiyoubi** [comment]
Then I guess it would be better to also change the video to avoid confusing other students:) 
I guess in Lecture5 24:07, remove all the **I** (Identity matrix) and change the vector `(x1;x2;x3;x4)` as `(x1^T; x2^T; x3^T; x4^T)` ?
